File: AccuPay/Domain/Payroll/ThirteenthMonthPayCalculator.vb
Option Strict On

Imports AccuPay.Entity
Imports AccuPay.Utilities
Imports PayrollSys

Namespace Global.AccuPay.Payroll

    Public Class ThirteenthMonthPayCalculator

        Public Sub Calculate(employee As Employee,
                           paystub As Paystub,
                           timeEntries As ICollection(Of TimeEntry),
                           actualtimeentries As ICollection(Of ActualTimeEntry),
                           salary As Salary,
                           settings As ListOfValueCollection,
                           allowanceItems As ICollection(Of AllowanceItem))

            If paystub.ThirteenthMonthPay Is Nothing Then
                paystub.ThirteenthMonthPay = New ThirteenthMonthPay() With {
                    .OrganizationID = z_OrganizationID,
                    .CreatedBy = z_User
                }
            Else
                paystub.ThirteenthMonthPay.LastUpdBy = z_User
            End If

            Dim thirteenthMonthAmount As Decimal

            Dim thirteenMonthCalculation = settings.GetEnum("ThirteenthMonthPolicy.CalculationBasis", ThirteenthMonthCalculationBasis.RegularPayAndAllowance)

            thirteenthMonthAmount = GetThirteenMonthAmount(paystub, thirteenMonthCalculation, employee, timeEntries, actualtimeentries, salary, settings, allowanceItems)

            paystub.ThirteenthMonthPay.BasicPay = thirteenthMonthAmount
            paystub.ThirteenthMonthPay.Amount = thirteenthMonthAmount / CalendarConstants.MonthsInAYear
            paystub.ThirteenthMonthPay.Paystub = paystub
        End Sub

        Private Shared Function GetThirteenMonthAmount(paystub As Paystub, thirteenMonthPolicy As ThirteenthMonthCalculationBasis, employee As Employee, timeEntries As ICollection(Of TimeEntry), actualtimeentries As ICollection(Of ActualTimeEntry), salary As Salary, settings As ListOfValueCollection, allowanceItems As ICollection(Of AllowanceItem)) As Decimal

            Select Case thirteenMonthPolicy
                Case ThirteenthMonthCalculationBasis.RegularPayAndAllowance

                    Return ComputeRegularPayAndAllowance(employee, timeEntries, actualtimeentries, salary, settings, allowanceItems)

                Case ThirteenthMonthCalculationBasis.DailyRate

                    Dim hoursWorked = paystub.TotalWorkedHoursWithoutOvertimeAndLeave

                    If (New SystemOwner).CurrentSystemOwner = SystemOwner.Benchmark AndAlso employee.IsPremiumInclusive Then

                        hoursWorked = paystub.RegularHoursAndTotalRestDay
                    End If

                    Dim daysWorked = hoursWorked / PayrollTools.WorkHoursPerDay

                    Dim dailyRate = PayrollTools.GetDailyRate(salary, employee)

                    Return AccuMath.CommercialRound(daysWorked * dailyRate)

                Case Else

                    Return 0

            End Select

        End Function

        Private Shared Function ComputeRegularPayAndAllowance(_employee As Employee, timeEntries As ICollection(Of TimeEntry), actualtimeentries As ICollection(Of ActualTimeEntry), salary As Salary, settings As ListOfValueCollection, allowanceItems As ICollection(Of AllowanceItem)) As Decimal
            Dim contractualEmployementStatuses = New String() {"Contractual", "SERVICE CONTRACT"}

            Dim thirteenthMonthAmount = 0D

            If _employee.IsDaily Then

                If contractualEmployementStatuses.Contains(_employee.EmploymentStatus) Then

                    thirteenthMonthAmount = timeEntries.
                                                Where(Function(t) Not t.IsRestDay).
                                                Sum(Function(t) t.BasicDayPay + t.LeavePay)
                Else

                    Dim thirteenthMonthAmountRunningTotal As Decimal = 0

                    For Each actualTimeEntry In actualtimeentries

                        Dim timeEntry = timeEntries.Where(Function(t) t.Date = actualTimeEntry.Date).FirstOrDefault

                        If timeEntry Is Nothing OrElse timeEntry.IsRestDay Then Continue For

                        thirteenthMonthAmount += actualTimeEntry.BasicDayPay + actualTimeEntry.LeavePay

                    Next

                End If

            ElseIf _employee.IsMonthly Or _employee.IsFixed Then

                Dim trueSalary = salary.TotalSalary
                Dim basicPay = trueSalary / CalendarConstants.SemiMonthlyPayPeriodsPerMonth

                Dim totalDeductions = actualtimeentries.Sum(Function(t) t.LateDeduction + t.UndertimeDeduction + t.AbsentDeduction)

                Dim additionalAmount As Decimal
                If (settings.GetBoolean("ThirteenthMonthPolicy.IsAllowancePaid")) Then

                    additionalAmount = allowanceItems.Sum(Function(a) a.Amount)

                End If

                thirteenthMonthAmount = ((basicPay + additionalAmount) - totalDeductions)
            End If

            Dim allowanceAmount = allowanceItems.Where(Function(a) a.IsThirteenthMonthPay).Sum(Function(a) a.Amount)
            thirteenthMonthAmount += allowanceAmount
            Return thirteenthMonthAmount
        End Function

    End Class

End Namespace

File: Accupay.Data/Services/PayrollGeneration/ThirteenthMonthPayCalculator.cs
using AccuPay.Data.Entities;
using AccuPay.Data.Enums;
using AccuPay.Data.Helpers;
using AccuPay.Utilities;
using System;
using System.Collections.Generic;
using System.Linq;

namespace AccuPay.Data.Services
{
    public class ThirteenthMonthPayCalculator
    {
        private readonly int _organizationId;
        private readonly int _userId;

        public ThirteenthMonthPayCalculator(int organizationId, int userId)
        {
            _organizationId = organizationId;
            _userId = userId;
        }

        public ThirteenthMonthPay Calculate(Employee employee,
                            Paystub paystub,
                            IReadOnlyCollection<TimeEntry> timeEntries,
                            IReadOnlyCollection<ActualTimeEntry> actualTimeEntries,
                            Salary salary,
                            ListOfValueCollection settings,
                            ICollection<AllowanceItem> allowanceItems,
                            string currentSystemOwner)
        {
            if (paystub.ThirteenthMonthPay == null)
                paystub.ThirteenthMonthPay = new ThirteenthMonthPay()
                {
                    OrganizationID = _organizationId,
                    CreatedBy = _userId,
                    Created = DateTime.Now
                };
            else
                paystub.ThirteenthMonthPay.LastUpdBy = _userId;

            decimal thirteenthMonthBasicPay;

            var thirteenMonthCalculation = settings
                .GetEnum("ThirteenthMonthPolicy.CalculationBasis", ThirteenthMonthCalculationBasis.RegularPayAndAllowance);

            thirteenthMonthBasicPay = GetThirteenMonthBasicPay(
                paystub,
                thirteenMonthCalculation,
                employee,
                timeEntries,
                actualTimeEntries,
                salary,
                settings,
                allowanceItems.ToList(),
                currentSystemOwner);

            var thirteenthMonthAmount = thirteenthMonthBasicPay / CalendarConstant.MonthsInAYear;

            paystub.ThirteenthMonthPay.BasicPay = thirteenthMonthBasicPay;
            paystub.ThirteenthMonthPay.Amount = thirteenthMonthAmount;
            paystub.ThirteenthMonthPay.Paystub = paystub;

            return paystub.ThirteenthMonthPay;
        }

        private static decimal GetThirteenMonthBasicPay(
            Paystub paystub,
            ThirteenthMonthCalculationBasis thirteenMonthPolicy,
            Employee employee,
            IReadOnlyCollection<TimeEntry> timeEntries,
            IReadOnlyCollection<ActualTimeEntry> actualtimeentries,
            Salary salary,
            ListOfValueCollection settings,
            IReadOnlyCollection<AllowanceItem> allowanceItems,
            string currentSystemOwner)
        {
            switch (thirteenMonthPolicy)
            {
                case ThirteenthMonthCalculationBasis.RegularPayAndAllowance:
                    return ComputeByRegularPayAndAllowance(employee, timeEntries, actualtimeentries, salary, settings, allowanceItems);

                case ThirteenthMonthCalculationBasis.DailyRate:
                    var hoursWorked = paystub.TotalWorkedHoursWithoutOvertimeAndLeave;

                    if (currentSystemOwner == SystemOwnerService.Benchmark && employee.IsPremiumInclusive)
                    {
                        hoursWorked = paystub.RegularHoursAndTotalRestDay;
                    }

                    return ComputeByDailyRate(employee, salary, hoursWorked);

                default:
                    return 0;
            }
        }

        private static decimal ComputeByDailyRate(Employee employee, Salary salary, decimal hoursWorked)
        {
            var daysWorked = hoursWorked / PayrollTools.WorkHoursPerDay;

            var dailyRate = PayrollTools.GetDailyRate(salary, employee);

            return AccuMath.CommercialRound(daysWorked * dailyRate);
        }

        private static decimal ComputeByRegularPayAndAllowance(
            Employee employee,
            IReadOnlyCollection<TimeEntry> timeEntries,
            IReadOnlyCollection<ActualTimeEntry> actualtimeentries,
            Salary salary,
            ListOfValueCollection settings,
            IReadOnlyCollection<AllowanceItem> allowanceItems)
        {
            var thirteenthMonthAmount = 0M;

            if (employee.IsDaily)
            {
                thirteenthMonthAmount = ComputeByRegularPayAndAllowanceDaily(
                    employee,
                    timeEntries,
                    actualtimeentries);
            }
            else if (employee.IsMonthly || employee.IsFixed)
            {
                if (salary == null) return 0;

                var trueSalary = salary.TotalSalary;
                var basicPay = trueSalary / CalendarConstant.SemiMonthlyPayPeriodsPerMonth;

                var totalDeductions = actualtimeentries.Sum(t => t.LateDeduction + t.UndertimeDeduction + t.AbsentDeduction);

                decimal additionalAmount = 0;
                if ((settings.GetBoolean("ThirteenthMonthPolicy.IsAllowancePaid")))
                {
                    // all allowance
                    additionalAmount = allowanceItems.Sum(a => a.Amount);
                }
                else
                {
                    // allowance that the 13th month pay flag is checked
                    additionalAmount = allowanceItems.Where(a => a.IsThirteenthMonthPay).Sum(a => a.Amount);
                }

                thirteenthMonthAmount = ((basicPay + additionalAmount) - totalDeductions);
            }

            return thirteenthMonthAmount;
        }

        // turn this back to private when we can use this class without coupling it with paystub
        // this should only be a calculator class, not an update paystub class
        public static decimal ComputeByRegularPayAndAllowanceDaily(
            Employee employee,
            IReadOnlyCollection<TimeEntry> timeEntries,
            IReadOnlyCollection<ActualTimeEntry> actualtimeentries)
        {
            decimal thirteenthMonthAmount = 0M;

            var contractualEmployementStatuses = new string[] { "Contractual", "SERVICE CONTRACT" };

            if (contractualEmployementStatuses.Contains(employee.EmploymentStatus))
            {
                thirteenthMonthAmount = timeEntries
                    .Where(t => !t.IsRestDay)
                    .Sum(t => t.BasicDayPay + t.LeavePay);
            }
            else
            {
                foreach (var actualTimeEntry in actualtimeentries)
                {
                    var timeEntry = timeEntries
                        .Where(t => t.Date == actualTimeEntry.Date)
                        .FirstOrDefault();

                    if (timeEntry == null || timeEntry.IsRestDay)
                    {
                        continue;
                    }

                    thirteenthMonthAmount += actualTimeEntry.BasicDayPay + actualTimeEntry.LeavePay;
                }
            }

            return thirteenthMonthAmount;
        }
    }
}